Authentication support for remote charts repositories

eyalbe4 opened the pull request:
This pull request adds support for sending credentials (username and password) to remote charts repositories. This will allow vendors of charts repositories to restrict access to privileged users only.
This new functionality is optional of course.

Please let us know what you think.
We'd be happy to add more commits as needed.
Thank you in advance.

eyalbe4 committed:
Authentication support for remote charts repositories.

eyalbe4 committed:
Authentication support for remote charts repositories.

eyalbe4 committed:
Authentication support for remote charts repositories.

eyalbe4 committed:
Authentication support for remote charts repositories.

eyalbe4 committed:
Authentication support for remote charts repositories.

eyalbe4 committed:
Authentication support for remote charts repositories.

eyalbe4 committed:
Tests fixes.

eyalbe4 committed:
Tests fixes.

eyalbe4 committed:
Ran 'go fmt' and 'make docs'.

eyalbe4 committed:
Ran 'go fmt' and 'make docs'.

eyalbe4 commented:
My CLA appears as unsigned, probably because I signed with a different email address (eyalbenmoshe@jfrog.com) than my github user's email (eyalb@jfrog.com).
Since I'm unable to sign the CLA again, I sent an email to the CNCF helpdesk, asking for help.

bacongobbler commented:
Your local git config is set to an incorrect email address when they're being signed.

```
commit 3e84ba1940b0e42e988433a8b2836e2a077c9019 (HEAD, eyalbe4/master)
Author: eyalbe4 <00Rotshild00>
Date:   Mon Nov 27 22:49:06 2017 +0200

    Ran 'go fmt' and 'make docs'.
```

You can change that with `git config --global user.email eyalb@jfrog.com`. and fix them up using `git remote add upstream https://github.com/kubernetes/helm && git fetch upstream master && git rebase -i upstream/master`.

eyalbe4 commented:
Thanks you for your help @bacongobbler!
All looks okay now.

eyalbe4 committed:
Bug fix - if --username and --password are not passed, use the configured repo credentials

eyalbe4 committed:
Bug fix - if --username and --password are not passed, use the configured repo credentials

eyalbe4 commented:
@bacongobbler,
We've been using the authentication mechanism in this PR for a few weeks already and it works great for us. 
Any chance this can be merged? Please let us know if you'd like us to add or change anything.
Thanks

thockin commented:
Who is approver for this?  @michelleN ?  @erictune ?

bacongobbler commented:
Anyone from helm's [OWNERS](https://github.com/kubernetes/helm/blob/master/OWNERS) file can review/approve. I can take a look.

bacongobbler commented on pkg/downloader/chart_downloader.go:
```diff
@@ -80,13 +80,24 @@ type ChartDownloader struct {
 // (if provenance was verified), or an error if something bad happened.
-func (c *ChartDownloader) DownloadTo(ref, version, dest string) (string, *provenance.Verification, error) {
-	u, g, err := c.ResolveChartVersion(ref, version)
+func (c *ChartDownloader) DownloadTo(ref, username, password, version, dest string) (string, *provenance.Verification, error) {
```
I feel like these should be fields of the ChartDownloader rather than fields in DownloadTo. That way the functions are backwards compatible for third party clients using pkg/downloader in their projects.

eyalbe4 replied:
Done.

bacongobbler commented on pkg/downloader/chart_downloader.go:
```diff
@@ -139,15 +150,15 @@ func (c *ChartDownloader) DownloadTo(ref, version, dest string) (string, *proven
 //		* If version is empty, this will return the URL for the latest version
 //		* If no version can be found, an error is returned
-func (c *ChartDownloader) ResolveChartVersion(ref, version string) (*url.URL, getter.Getter, error) {
+func (c *ChartDownloader) ResolveChartVersion(ref, version string) (*url.URL, getter.Getter, *repo.Entry, error) {
```
If we're going to change this function, we should just refactor it to return the ChartRepository (`r`) instead of `r.Client` and `r.Entry` separately. It's backwards-incompatible, but third parties can still receive r.Client in their code. It would make this code a *lot* cleaner by not having to return `nil, nil` or `r.Client, r.Config` everywhere.

eyalbe4 replied:
Done.

eyalbe4 committed:
Basic auth support - Refactor.

eyalbe4 committed:
Basic auth support - Refactor.

eyalbe4 commented:
Thanks for reviewing the code @bacongobbler.
I implemented both comments.
Please let me know if additional changes are needed.

eyalbe4 commented:
@bacongobbler / helm [OWNERS](https://github.com/kubernetes/helm/blob/master/OWNERS),
Do you think this PR can be merged?

eyalbe4 commented:
@prydonius, @erictune, @adamreese, @SlickNik, @michelleN,
Is there anything else I need to do to get this functionality merged?

bacongobbler commented:
Would it be too much to ask for formal documentation under docs/? :)

The functionality is all there there, but how one actually sets up a chart repository and how the authentication mechanism is handled from the end-user's point of view would be nice to see. I'm not sure how to go about and test or use this new feature.

As a side note, how does this affect the existing feature set such as https://github.com/kubernetes/helm/pull/1710?

eyalbe4 committed:
Documentation and tests.

eyalbe4 committed:
Documentation and tests.

eyalbe4 commented:
@bacongobbler,
I've pushed another commit adding some documentation and tests.
Please let me know if the following answers your questions. I'd be glad to add more  information:

[Artifactory](https://www.jfrog.com/confluence/display/RTF/Helm+Chart+Repositories) recently started supporting chart repositories. Since Artifactory is mostly accessed using basic auth, adding basic auth support to helm seems necessary when working with a remote chart repository provider, such as Artifactory.

As *chart_downloader_test.go* shows, the existing helm client does support basic auth, but only by adding the credentials as part of the URL. This however means that the URL (with the credentials in it) should be used in the *requirements.yaml*. This pull request however completely resolves this issue, by saving the credentials separately.

As for the basic auth test in https://github.com/kubernetes/helm/pull/1710, I slightly modified it, so that it uses the new getter.GetWithCredentials API. The modified test works great.
I also fixed a small bug in the test:
The test used to fail only if both the username and password sent were incorrect. I fixed it, so that it fails if either of them is incorrect.

I also added some documentation to *helm/docs/chart_repository.md*.
Please let me know if you'd like me to modify it or add some more documentation.

Thanks

eyalbe4 commented:
@bacongobbler,
Does my latest response and commit answer your questions?
Is there anything else you'd like me add so that we can get this merged?

eldada commented:
BTW - I've been using the forked version with credentials against Artifactory as my helm repository, and it's working as designed.

cmeury commented:
Not sure it helps the cause, but I also successfully pushed and pulled from Artifactory helm repository using the forked version maintained by JFrog.

thockin commented:
@bacongobbler  - Anything in particular blocking this?  It's been on my dashboard for quite a while...

bacongobbler commented:
Sorry, been blocked by higher priority work (helm 2.8.1, helm dev summit, etc). I’ll take a look at this tomorrow.

michelleN commented on docs/helm/helm_repo_add.md:
````diff
@@ -8,7 +8,7 @@ add a chart repository
 
 ```
-helm repo add [flags] [NAME] [URL]
+helm repo add [flags] [NAME] [URL] [USERNAME] [PASSWORD]
````
Would you mind making these flags instead of args? The general rule of thumb we use are `args should be required and flags are optional`. This will also make the experience consistent across the other commands that can take a username and password.

eyalbe4 replied:
I changed them to be flags.

michelleN commented on cmd/helm/fetch.go:
```diff
@@ -117,6 +119,8 @@ func (f *fetchCmd) run() error {
 		Keyring:  f.keyring,
 		Verify:   downloader.VerifyNever,
 		Getters:  getter.All(settings),
+		Username: f.username,
```
It looks like the `--username` and `--password` flags have not been added to this command. If those were left off by mistake, would you mind adding those in?

eyalbe4 replied:
You're right. They were left out by mistake.
I've just added them.

michelleN commented on docs/chart_repository.md:
````diff
@@ -271,7 +276,7 @@ If the charts are backed by HTTP basic authentication, you can also supply the
 
 ```console
-$ helm repo add fantastic-charts https://username:password@fantastic-charts.storage.googleapis.com
+$ helm repo add fantastic-charts https://username:password@fantastic-charts.storage.googleapis.com my-username my-password
````
This should probably be updated to `$ helm repo add fantastic-charts https://fantastic-charts.storage.googleapis.com --username my-username --password my-password` once the args are turned into flags.

michelleN commented on pkg/repo/chartrepo.go:
```diff
@@ -185,7 +188,7 @@ func (r *ChartRepository) generateIndex() error {
 // FindChartInRepoURL finds chart in chart repository pointed by repoURL
 // without adding repo to repositories
-func FindChartInRepoURL(repoURL, chartName, chartVersion, certFile, keyFile, caFile string, getters getter.Providers) (string, error) {
+func FindChartInRepoURL(repoURL, username, password, chartName, chartVersion, certFile, keyFile, caFile string, getters getter.Providers) (string, error) {
```
This is a breaking change since it is a change to a public function signature. We'll probably want to create a new function here that can handle taking in a username and password.

eyalbe4 replied:
Fixed.

michelleN reviewed (changes requested):
Hey @eyalbe4, 
Thanks so much for your patience as we go through this PR. I know it's been in the queue a while. Because this change touches so many files and is cross cutting, we'll want to get LGTMs from two core maintainers. Also why I am jumping in here! 👋 

I tested this PR manually going through the commands you've changed as well as others. I appreciate the tests you've added. Overall, it looks great. I did leave a few questions for you around consistency of experience, return values, and one backwards incompatible change I saw. I was really looking out for any backwards incompatible changes and I think this PR does a great job of keeping everything in tact. There is just one function in question.

Once we have those items resolved, I'd feel comfortable moving forward. Please feel free to ping me on slack (@mnoorali) if you need. Thank you again for your patience and for contributing.

-------------
P.S. One other (non-blocking) thing I noticed was that the username and password flags had to be set on every command if you're working with a chart repo that has basic auth enabled. In a follow up PR, it'd be great to not have to see if we can leverage the cached username/pw info in `~/.helm/repository/repositories.yaml` so that a user doesn't have to pass in the flags every time. Of course, that should be in a separate issue as a follow up item with more discussion with others.

eyalbe4 commented on pkg/downloader/manager.go:
```diff
@@ -459,22 +461,30 @@ func (m *Manager) parallelRepoUpdate(repos []*repo.Entry) error {
+			entry, err = findEntryByName(name, cr)
 			if err != nil {
-				return "", err
+				return
```
Since the *findChartURL* function now returns named parameters, there's actually no need to specify the returned values. Before my change, the method returned:
```
(string, error)
```
Now it returns:
```
(url, username, password string, err error)
```
So the value of *err* is returned anyway. 
Please let me know if you'd like me to change this.

michelleN replied:
Oh I didn't catch that. Thank you!

eyalbe4 commented on pkg/downloader/manager.go:
```diff
@@ -459,22 +461,30 @@ func (m *Manager) parallelRepoUpdate(repos []*repo.Entry) error {
 	}
-	return "", fmt.Errorf("chart %s not found in %s", name, repoURL)
+	err = fmt.Errorf("chart %s not found in %s", name, repoURL)
+	return
```
Same as my response your previous comment, the *findChartURL* now uses named returned values, so *err* is returned without adding it to the return statement.
Please let me know if you'd like me to change this.

michelleN replied:
No that's great. Thanks

eyalbe4 committed:
From args to flags, fetch command fix and backward compatibility fix.

eyalbe4 committed:
From args to flags, fetch command fix and backward compatibility fix.

eyalbe4 committed:
helm docs update.

eyalbe4 committed:
helm docs update.

eyalbe4 commented:
Hi @michelleN,

Thank you for the feedback and code review! 
I pushed a commit fixing the issues you mentioned. 
As for the questions you asked about the * findChartURL* function, please see my replies, and let me know if you'd like me to make the changes you proposed.

If you'd like to leverage the username and password in a future pull request, please let me know when the time comes and I'll gladly contribute the code.

Thanks again for spending the time and helping push this important functionality further. I really appreciate it.
Please let me know if there are other changes you'd like to see, and I'll add them.

michelleN commented on pkg/downloader/chart_downloader.go:
```diff
@@ -139,7 +156,7 @@ func (c *ChartDownloader) DownloadTo(ref, version, dest string) (string, *proven
 //		* If version is empty, this will return the URL for the latest version
 //		* If no version can be found, an error is returned
-func (c *ChartDownloader) ResolveChartVersion(ref, version string) (*url.URL, getter.Getter, error) {
+func (c *ChartDownloader) ResolveChartVersion(ref, version string) (*url.URL, *repo.ChartRepository, error) {
```
Looks like this will be a backwards incompatible change because we're changing the return values.

eyalbe4 replied:
Fixed: 
The old signature of the *ResolveChartVersion* is back.
I added a new method - *ResolveChartVersionAndGetRepo*, with the new returned values.

michelleN commented on pkg/repo/chartrepo.go:
```diff
@@ -186,6 +189,13 @@ func (r *ChartRepository) generateIndex() error {
 // FindChartInRepoURL finds chart in chart repository pointed by repoURL
 // without adding repo to repositories
 func FindChartInRepoURL(repoURL, chartName, chartVersion, certFile, keyFile, caFile string, getters getter.Providers) (string, error) {
+	return FindChartInAuthRepoURL(repoURL, "", "", chartName, chartVersion, certFile, keyFile, caFile, getters)
```
👍

michelleN reviewed (changes requested):
Thanks @eyalbe4! I took it for another spin and it's working great. Got a few more questions for you.
- I noticed that if I add a repo with a `username` and `password` then do a `helm fetch/inspect/etc name/chart` without a `username` and `password` flag, it will return successfully (fetch/inspect a chart without having to pass creds in). 
Can you verify that is expected behavior?
- I noticed a possible backwards incompatible change <noted in second review>. If you could address that, I'd appreciate it.
- Would you mind regenerating the docs? It's looking like the new username password flags on helm fetch haven't been updated.

Thanks again!

bacongobbler commented on docs/chart_repository.md:
````diff
@@ -270,8 +275,8 @@ fantastic-charts    https://fantastic-charts.storage.googleapis.com
-```console
-$ helm repo add fantastic-charts https://username:password@fantastic-charts.storage.googleapis.com
+``console
+$ helm repo add fantastic-charts https://fantastic-charts.storage.googleapis.com --username my-username--password my-password
````
missing whitespace between --password and the value supplied to --username.

eyalbe4 replied:
Thanks for noticing this. Fixed.

eyalbe4 committed:
Backwards compatibility and doc fix.

eyalbe4 committed:
Backwards compatibility and doc fix.

eyalbe4 committed:
helm-fetch.md update.

eyalbe4 committed:
helm-fetch.md update.

eyalbe4 commented:
@michelleN,
Thanks for reviewing the code again. I changed the code according to your suggestions:
I regenerated the docs and committed the updated *helm-fetch.md*.
I also add a new method signature, so that the old method signature remains as it was.

As for the usage of the username and password by the various commands, I think that the behaviour you experienced is expected. Here's why: 
All commands take the username and password flags values (if sent) or the configured username and password (if configured) and send them to the remote chart repository, by adding them as a basic auth header. If the chart repository however is not configured to validate the credentials, it will just ignore the header and the request should not fail. Please let me know if you have more questions about this, and I'd be happy to elaborate.

I can see that there are a few conflicting files. Since all of them are under helm/docs, they can be easily resolved once this branch is synched with the latest code. I thought is might be easier to do this once, when the code in this PR is approved. If you however prefer I do the sync now, please let me know.

Thanks again @michelleN and @bacongobbler for the review and please let me know if there's anything else you'd like to see.

michelleN commented:
Thanks @eyalbe4 I did another round of manual tests and read through the code and it all looks great to me. Please resolve the merge conflicts so we can merge after @bacongobbler has had a chance to review!

bacongobbler commented on pkg/getter/plugingetter.go:
```diff
@@ -79,6 +79,10 @@ func (p *pluginGetter) Get(href string) (*bytes.Buffer, error) {
 }
 
+func (p *pluginGetter) GetWithCredentials(href, username, password string) (*bytes.Buffer, error) {
+	return p.Get(href)
```
This is the reason for the above comment... This is a very odd design choice. Why should Getters provide a `GetWithCredentials` function if they aren't going to use the credentials provided?

eyalbe4 replied:
@bacongobbler,
If understand your comment correctly, you're saying that only the *getter.httpGetter* struct actually needs credentials.  This will probably require the following the change:
The *ChartRepository* struct currently holds a *getter.Getter*. We'll need to replace it with *getter.httpGetter*. We need this because the *ChartDownloader.DownloadTo* method uses the getter from the *ChartRepository* struct, to download with the credentials.
Does this makes sense? Do you think it is okay to have the *ChartRepository* hold *getter.httpGetter* instead of a generic *getter.Getter*?
If so, I'll make this change.

bacongobbler replied:
I do not think that it's okay to change the ChartRepository to use a `getter.HTTPGetter`. My current thinking is that when `ResolveChartVersionAndGetRepo` is called, it should be returning a `getter.HTTPGetter` with the username and password configuration as part of the struct, essentially moving [this entire code of logic](https://github.com/eyalbe4/helm/blob/50a3306a72830e1308c37437866d7abee0173d09/pkg/downloader/chart_downloader.go#L93-L104) into `ResolveChartVersionAndGetRepo`.

Does that help make sense?

eyalbe4 replied:
Sure @bacongobbler. Done.

bacongobbler commented on pkg/getter/getter.go:
```diff
@@ -27,6 +27,8 @@ import (
 	//Get file content by url string
 	Get(url string) (*bytes.Buffer, error)
+	//Get file content by url, username and password strings
+	GetWithCredentials(href, username, password string) (*bytes.Buffer, error)
```
Let's remove this. A Getter should just implement `.Get()` with a URL. If a username/password is provided as part of the struct that implements the Getter interface, it can then call `getWithCredentials`.

eyalbe4 replied:
Done.

bacongobbler reviewed (changes requested):
I have a few concerns with the how `pkg/` has been changed and/or architected with this change, but from a practical point of view the PR works as described. Once the comments have been addressed and the PR rebased this looks like it's ready to :shipit:!

In regards to [my previous comment](https://github.com/kubernetes/helm/pull/3206#issuecomment-358885158), I still believe it would be useful if there was more documentation around how the Helm client interacts with chart repositories backed by basic auth. For example, @michelleN and I were using https://github.com/michelleN/secure-chart-repo to test this PR *instead of artifactory* to make sure that this PR was vendor neutral and others like @jdolitsky could take advantage of this new feature in projects like ChartMuseum.

eyalbe4 commented:
@bacongobbler,
I rebased the commits and resolved the conflicts.
As for your comments, see my inline questions and let me know what you think.
Thanks

eyalbe4 commented:
@bacongobbler,
Can you please take a look at my question following your requested change and let me know how you'd like me to proceed?

thockin commented:
This PR is only interesting to me in that I am trying to figure out how to NOT let PRs languish for months on end.  @bacongobbler do you actually have time for it?

bacongobbler commented:
We were just at Helm summit last week; I will take a look again tomorrow.

bacongobbler commented:
> This PR is only interesting to me in that I am trying to figure out how to NOT let PRs languish for months on end.

OK, I feel like I need to explain why this PR has been sitting in the queue for a long period of time. The short form is that both parties here (being myself and @eyalbe4) are equally to blame.

Let's start with my end first.

When first triaging this issue, this PR did not have any discussion in the dev calls, there were no bugs referenced, and no documentation. That made it *incredibly* difficult to triage, test and/or assess the impact this feature had on the project. What I should have done is immediately ask for clarification on the PR, which I didn't do until [several months later](https://github.com/kubernetes/helm/pull/3206#issuecomment-358885158). I'll take the blame on that front and learn from this for future discussions.

Now, on @eyalbe4's end...

When this PR was originally posted, there was no discussion, referenced bugs in the OP or documentation in any shape or form. That made this PR nearly impossible to test because there were no existing chart repositories to test against asides JFrog's artifactory. I'll note that Artifactory was referenced in a follow-up commit, but it was unclear how someone could utilize this new feature against Artifactory, as well as why it was better than the previous form (again, we didn't ask for clarification [until several months later](https://github.com/kubernetes/helm/pull/3206#issuecomment-359267577)).

From @eyalbe4's end, it would have been *much* more useful if there was a documented list of issues this PR solved, a set of steps to test this PR with Artifactory or a reference architecture we could test with. Through our own frustration, we gave up testing with Artifactory and @michelleN made [her own mock chart repository](https://github.com/michelleN/secure-chart-repo) to test this PR against which worked for demonstration and testing purposes.

Hopefully that clarifies the situation a bit. Now, let's get back to actually moving this PR forward.

thockin commented:
To be clear: no blame.  Just want to try to make it better for people.
Would be good to augment some of the contributor docs to explain the
best-practices for streamlining a PR.

On Tue, Feb 27, 2018 at 8:36 AM, Matthew Fisher <notifications@github.com>
wrote:

> This PR is only interesting to me in that I am trying to figure out how to
> NOT let PRs languish for months on end.
>
> OK, I feel like I need to explain why this PR has been sitting in the
> queue for a long period of time. The short form is that both parties here
> (being myself and @eyalbe4 <https://github.com/eyalbe4>) are equally to
> blame.
>
> Let's start with my end first.
>
> When first triaging this issue, this PR did not have any discussion in the
> dev calls, there were no bugs referenced, and no documentation. That made
> it *incredibly* difficult to triage, test and/or assess the impact this
> feature had on the project. What I should have done is immediately ask for
> clarification on the PR, which I didn't do until several months later
> <https://github.com/kubernetes/helm/pull/3206#issuecomment-358885158>.
> I'll take the blame on that front and learn from this for future
> discussions.
>
> Now, on @eyalbe4 <https://github.com/eyalbe4>'s end...
>
> When this PR was originally posted, there was no discussion, referenced
> bugs in the OP or documentation in any shape or form. That made this PR
> nearly impossible to test because there were no existing chart repositories
> to test against asides JFrog's artifactory. I'll note that Artifactory was
> referenced in a follow-up commit, but it was unclear how someone could
> utilize this new feature against Artifactory, as well as why it was better
> than the previous form (again, we didn't ask for clarification until
> several months later
> <https://github.com/kubernetes/helm/pull/3206#issuecomment-359267577>).
>
> From @eyalbe4 <https://github.com/eyalbe4>'s end, it would have been
> *much* more useful if there was a documented list of issues this PR
> solved, a set of steps to test this PR with Artifactory or a reference
> architecture we could test with. Through our own frustration, we gave up
> testing with Artifactory and @michelleN <https://github.com/michellen>
> made her own mock chart repository
> <https://github.com/michelleN/secure-chart-repo> to test this PR against
> which worked for demonstration and testing purposes.
>
> Hopefully that clarifies the situation a bit. Now, let's get back to
> actually moving this PR forward.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/helm/pull/3206#issuecomment-368940767>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVP9GiI2Ja3JcO6NuhbnPBfPJ3avVks5tZC8dgaJpZM4QsJ2n>
> .
>

eyalbe4 committed:
Removed GetWithCredentials API.

eyalbe4 committed:
Removed GetWithCredentials API.

eyalbe4 commented:
@bacongobbler,

Thank you for sharing these guidelines for requesting new functionalities and submitting pull requests. I wasn't aware that testing this can be challenging. Since servers which allow anonymous access are not affected by this, all we needed to test, was that the basic auth header is passed for every command (in addition for the commands functionality of course).
Since this functionality can be used by any future charts repo vendor (we tested it with Artifactory before every commit of course), I didn't think that you'd be interested in testing it specifically with Artifactory.
I will make sure to follow all of these guidelines for future requests and contributions.

I implemented the changes you requested and removed the *GetWithCredentials* API. This change required the following changes which I also implemented:
* *getter.httpGetter* was renamed to *getter.HttpGetter*, because it now needs to be available also outside of the *getter* package.
* The *ResolveChartVersionAndGetRepo* method now also returns a *getter.HttpGetter* argument.
* A new *getter.NewHTTPGetter* function has been added (in addition to the existing *getter.newHTTPGetter* function), to allow the *ResolveChartVersionAndGetRepo* to obtain and return a *getter.HttpGetter* argument.

Looking forward to your review and feedback. Thanks.

eyalbe4 commented:
@bacongobbler,
If we manage to reduce our response time while working on this, we should be able to finish the work quickly.
Please LMK what you think.

mfilotto commented:
@bacongobbler any update on this one ? Thanks for your help

bacongobbler reviewed (approved):
Looks good to merge once it's been rebased. Thanks!

mfilotto commented:
@eyalbe4 ok for you to rebase ?
Sorry guys but I really need this feature :-)

eyalbe4 commented:
Absolutely @mfilotto.
I'll rebase this today.

adamreese committed:
fix(helm): fix helm history unit tests

fixes #3652

eyalbe4 committed:
Fixed inspect.go followint rebase.

eyalbe4 committed:
Fixed inspect.go followint rebase.

eyalbe4 commented:
@bacongobbler,
The code has been rebased, but the tests are failing, probably as a result of a different commit added to the master:
```
# k8s.io/helm/cmd/helm
./history_test.go:61:8: unknown field 'cmds' in struct literal of type releaseCase
./history_test.go:62:8: unknown field 'desc' in struct literal of type releaseCase
./history_test.go:64:8: cannot use []*release.Release literal (type []*release.Release) as type *release.Release in field value
./history_test.go:68:8: unknown field 'xout' in struct literal of type releaseCase
./history_test.go:71:8: unknown field 'cmds' in struct literal of type releaseCase
./history_test.go:72:8: unknown field 'desc' in struct literal of type releaseCase
./history_test.go:74:8: cannot use []*release.Release literal (type []*release.Release) as type *release.Release in field value
./history_test.go:78:8: unknown field 'xout' in struct literal of type releaseCase
FAIL	k8s.io/helm/cmd/helm [build failed]
```

cmeury commented:
@eyalbe4 Are you sure all the commits should show up in this branch, this seems unusual? I would normally do a git fetch first, update my local master branch, change to the local feature branch and do:

    git rebase master
    git push -f origin foo:foo

This should lead to a branch/PR to only contain the actual changes, exactly as the branch would be created based on recent master.

bacongobbler commented:
See https://github.com/kubernetes/helm/pull/3717

andriisoldatenko committed:
Fixed typo

eyalbe4 commented:
@cmeury,
You're probably right and I could have been smarter about how I did the rebase...
I think we're good though. All of this PR's commits are now at the top. [Fixed inspect.go following rebase.](https://github.com/kubernetes/helm/pull/3206/commits/d0e3561ada592aa3a39a7a3bc255689ffc2c194a) is the only new commit I had to add following the rebase.
I hope that after https://github.com/kubernetes/helm/pull/3717 is committed, we'll be good to go.

Commit:
Merge pull request #3719 from andriisoldatenko/main/typo-in-comments

Fixed typo

adamreese committed:
Merge pull request #3717 from adamreese/fix/history-output

fix(helm): fix helm history unit tests

eyalbe4 committed:
Merge branch 'master' of https://github.com/eyalbe4/helm

eyalbe4 commented:
Tests look good now.

seriousben reviewed (approved):
🎉

kopf commented:
Congrats! Very glad to see this merged. Thanks to all involved. 

I held off with commenting as I didn't want to derail your discussions. 

The Artifactory set-up I'm using also requires a client-side SSL cert. When I noticed that this PR just adds support for Basic Auth, I had a look at the diff to see how much work it'd be to add support for client-side ssl certs. By the looks of things, there's no centralised "HTTP client configuration" or similar - instead the username/password are passed around throughout the code. 

So, I'm wondering - if I were to add client-side SSL cert support, would it be OK to do it in the way that was done in this PR, or would the maintainers (ping @bacongobbler @michelleN) feel that this is the time to do a more involved refactoring and store this HTTP client configuration somewhere centralised?

The reason I ask is that I have absolutely 0 Go experience, but have been wanting to get into it for some time. I feel this would be a nice starter project, and could easily get it done by emulating the approach in this PR. If a big refactoring was to be done, however, I'd perhaps need someone to dictate the design a little.

jgiles commented:
@kopf I think that client-side TLS certs are already supported in the Helm client?

https://github.com/kubernetes/helm/issues/1038#issuecomment-297043419

(As an aside, since I'm also about to set up Artifactory - is Artifactory itself doing the client-cert checking, or is it just behind some proxy that does the check? If Artifactory itself, how did you set that up?)

kopf commented:
@jgiles - ah, super - thanks for the heads-up!

I didn't set it up personally, but if I were going to do it, I'd personally use an nginx ssl-offloader.

bacongobbler commented:
I didn't set up Artifactory either when testing. I'd ask JFrog directly :)

mfilotto commented:
Did someone try to download dependencies from  an authenticated repository ?
I did a `helm dependency update` with no luck
I get an error ` 401 Unauthorized`
@eyalbe4 an idea ?
```
[root@dbk-miniont-01v nginx-ingress-sln]$ cat requirements.yaml
# requirements.yaml
dependencies:
 - name: consul
   version: "1.3.5"
   repository: "@stable"
```
```
[root@dbk-miniont-01v nginx-ingress-sln]$ helm repo list
NAME    URL
stable  https://artifactory.sln.nc/artifactory/helm
```
```
[root@dbk-miniont-01v nginx-ingress-sln]$ helm dependency update
Hang tight while we grab the latest from your chart repositories...
...Unable to get an update from the "stable" chart repository (https://artifactory.sln.nc/artifactory/helm):
        Failed to fetch https://artifactory.sln.nc/artifactory/helm/index.yaml : 401 Unauthorized
Update Complete. ⎈Happy Helming!⎈
Saving 1 charts
Downloading consul from repo https://artifactory.sln.nc/artifactory/helm
Save error occurred:  could not download https://artifactory.sln.nc:443/artifactory/helm/consul-1.3.5.tgz: Failed to fetch https://artifactory.sln.nc:443/artifactory/helm/consul-1.3.5.tgz : 401 Unauthorized
Deleting newly downloaded charts, restoring pre-update state
Error: could not download https://artifactory.sln.nc:443/artifactory/helm/consul-1.3.5.tgz: Failed to fetch https://artifactory.sln.nc:443/artifactory/helm/consul-1.3.5.tgz : 401 Unauthorized
```

bacongobbler commented:
Hey! Which version of helm are you running? This is only available in the canaries and 2.9.0-rc4 at this point.

mfilotto commented:
I use v2.9.0-rc4 with a repo added with username/password

eyalbe4 commented:
I'm looking into this @mfilotto.
In the meantime, can you look in Artifactory's request log (available through the UI) and let me know if the request sent by helm includes the user configured in helm?
Just in case we'll need to debug this on your end, which operating system are you running helm on?

eyalbe4 commented:
@mfilotto,
Artifactory currently supports updates from virtual repositories only.
Could it be that you're getting the *401 Unauthorised* because your repositories.yaml is not referencing a virtual repository? I've just tested this with a virtual repository and it seems to be working as expected (401 if my credentials are incorrect and a successful update if they are correct.
I did manage to get a 401 when I replaced it with a remote repository.

mfilotto commented:
@eyalbe4 Here is my config
- Artifactory : 5.10.1
- Helm : helm-v2.9.0-rc4-linux-amd64
- Helm OS: CentOS 7.1.1503

I use an helm virtual repository.
The search command list all availables charts : user auhtenticated
The dependencies update return 401: non_authenticated_user

```
20180426115727|1678|REQUEST|10.110.24.56|mathieu.filotto|GET|/helm/index.yaml|HTTP/1.0|200|1967664
20180426115729|0|REQUEST|10.110.24.56|non_authenticated_user|GET|/helm/consul-1.3.5.tgz|HTTP/1.0|401|0
```

mfilotto commented:
Helm version v2.9.0-rc5 solved my issue :smiley: 
:+1:
